### PR TITLE
#917: [typescript .d.ts] Badge (closes #917)

### LINKED ISSUE #917: [typescript .d.ts] Badge
← #908

## requirements
{describe the new feature}

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by ascariandrea**
Closes #917

### tests performed
- tested on `alinity-pro` on `ts-brc` branch ✅ 

- [ ] Google Chrome
- [ ] Internet Explorer

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/Divider/Divider.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+declare namespace Divider {
+
+  export type Orientation = 'horizontal' | 'vertical';
+  export type Size = 'small' | 'medium' | 'large' | 'no-margin';
+
+  type Props = {
+    orientation: Orientation,
+    size: Size,
+    style: object
```
Comment (FrancescoCioria) [inline src/Divider/Divider.d.ts]: all these props should be optional (append a `?` to each key)

File: src/Divider/Divider.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  }
+}
+
+declare class Divider extends PureComponent<Divider.Props, void> {}
+
+export = Divider
```
Comment (FrancescoCioria) [inline src/Divider/Divider.d.ts]: mmm why isn't this an `export default` like the `js` file?

File: src/badge/Badge.d.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { PureComponent, ReactChildren } from 'react';
+
+declare namespace Badge {
+
+  type Props = {
+    label: ReactChildren,
```
Comment (FrancescoCioria) [inline src/badge/Badge.d.ts]: Typescript's `ReactChildren` is quite different from `tcomb`'s one
I'm pretty sure this is not enough...

I tried this quickly in a project and it complains:
```ts
type X = {
    child: React.ReactChildren
  };

    const x: X = {
      child: <div />
    };

    const x: X = {
      child: ''
    };
```

File: src/badge/Badge.js [hunk 4]
```diff
@@ -35,4 +34,8 @@ export default class Badge extends React.PureComponent {
       </FlexView>
     );
   }
+
+  render() {
+    return this.template(this.getLocals());
```
Comment (FrancescoCioria) [inline src/badge/Badge.js]: why this change? we're shouldn't touch `.js` files in these issues

File: src/button/Button.d.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    type?: ButtonType,
+    primary?: boolean,
+    flat?: boolean,
+    size?: ButtonSize,
+    fluid?: boolean,
+    circular?: boolean,
+    textOverflow?: React.ComponentClass<object>, // todo(typo) when translate text overflow, update here its props types
+    style?: object,
+    className?: string,
+    baseState?: ButtonBaseState,
```
Comment (FrancescoCioria) [inline src/button/Button.d.ts]: these last three props should be removed as they are from `StatefulButton`

File: src/button/StatefulButton.d.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { PureComponent } from 'react';
+import * as Button from './Button';
+
+declare namespace StatefulButton {
+  type Props = Button.Props;
```
Comment (FrancescoCioria) [inline src/button/StatefulButton.d.ts]: and they should be added here with an `&`:
```
type Props = Button.Props & { baseState?: ...}
```

File: src/text-overflow/TextOverflow.d.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { PureComponent } from 'react';
+
+declare namespace TextOverflow {
+  type Props = {
+    label?: string | number,
```
Comment (FrancescoCioria) [inline src/text-overflow/TextOverflow.d.ts]: `label` is required and should accept only `string`

File: src/text-overflow/TextOverflow.d.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    popover?: object, // TODO(typo): Omit?
+    lazy?: boolean,
+    id?: string,
+    className?: string,
+    style?: object
+  }
```
Comment (FrancescoCioria) [inline src/text-overflow/TextOverflow.d.ts]: you should reflect exactly the `Props` object from `TextOverflow.js`

here you added `popover` while `children` and `delayWhenLazy` are both missing.

File: src/button/Button.d.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    label?:string | object,
+    icon?:string | object,
+    children?: string,
+    type?: ButtonType,
+    primary?: boolean,
+    flat?: boolean,
+    size?: ButtonSize,
+    fluid?: boolean,
+    circular?: boolean,
+    textOverflow?: React.ComponentClass<object>, // todo(typo) when translate text overflow, update here its props types
```
Comment (FrancescoCioria) [inline src/button/Button.d.ts]: you did it in this PR, right? can't you use it here yet?

File: src/Divider/Divider.d.ts (hunk 2 above)
Comment (ascariandrea) [inline src/Divider/Divider.d.ts]: cause write as this let us to:
```ts
import * as Divider from 'path/to/divider';

class MyDivider extender React.PureComponent<Divider.Props, void> {
 render() {
  return <Divider {...this.props} />
 }
}
```
